stores/multipart.go

```diff
@@ -0,0 +1,341 @@
+		err := tx.Table("(SELECT id from buckets WHERE buckets.name = ?) bucket_id", bucket).
+			Take(&bucketID).Error
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return fmt.Errorf("bucket %v not found: %w", bucket, api.ErrBucketNotFound)
+		} else if err != nil {
+			return fmt.Errorf("failed to fetch bucket id: %w", err)
+		}
+		// Create multipart upload
+		uploadIDEntropy := frand.Entropy256()
+		uploadID = hex.EncodeToString(uploadIDEntropy[:])
+		if err := s.db.Create(&dbMultipartUpload{
+			DBBucketID: bucketID,
+			Key:        []byte(object.NoOpKey.String()), // TODO: set actual key
+			UploadID:   uploadID,
+			ObjectID:   path,
+		}).Error; err != nil {
+			return fmt.Errorf("failed to create multipart upload: %w", err)
+		}
+		return nil
+	})
+	return api.MultipartCreateResponse{
+		UploadID: uploadID,
+	}, err
+}
+
+func (s *SQLStore) AddMultipartPart(ctx context.Context, bucket, path, contractSet, uploadID string, partNumber int, slices []object.SlabSlice, partialSlabs []object.PartialSlab, etag string, usedContracts map[types.PublicKey]types.FileContractID) (err error) {
+	err = s.retryTransaction(func(tx *gorm.DB) error {
+		// Fetch contract set.
+		var cs dbContractSet
+		if err := tx.Take(&cs, "name = ?", contractSet).Error; err != nil {
+			return fmt.Errorf("contract set %v not found: %w", contractSet, err)
+		}
+		// Fetch the used contracts.
+		contracts, err := fetchUsedContracts(tx, usedContracts)
+		if err != nil {
+			return fmt.Errorf("failed to fetch used contracts: %w", err)
+		}
+		// Find multipart upload.
+		var mu dbMultipartUpload
+		err = tx.Where("upload_id", uploadID).
+			Take(&mu).
+			Error
+		if err != nil {
+			return fmt.Errorf("failed to fetch multipart upload: %w", err)
+		}
+		// Delete a potentially existing part.
+		err = tx.Model(&dbMultipartPart{}).
+			Where("db_multipart_upload_id = ? AND part_number = ?", mu.ID, partNumber).
+			Delete(&dbMultipartPart{}).
+			Error
+		if err != nil {
+			return fmt.Errorf("failed to delete existing part: %w", err)
+		}
+		var size uint64
+		for _, slice := range slices {
+			size += uint64(slice.Length)
+		}
+		for _, ps := range partialSlabs {
+			size += uint64(ps.Length)
+		}
+		// Create a new part.
+		part := dbMultipartPart{
+			Etag:                etag,
+			PartNumber:          partNumber,
+			DBMultipartUploadID: mu.ID,
+			Size:                size,
+		}
+		err = tx.Create(&part).Error
+		if err != nil {
+			return fmt.Errorf("failed to create part: %w", err)
+		}
+		// Create the slices.
+		err = s.createSlices(tx, nil, &part.ID, cs.ID, contracts, slices, partialSlabs)
+		if err != nil {
+			return fmt.Errorf("failed to create slices: %w", err)
+		}
+		return nil
+	})
+	return err
+}
+
+// TODO: f/u with support for 'prefix', 'keyMarker' and 'uploadIDMarker'
+func (s *SQLStore) ListMultipartUploads(ctx context.Context, bucket, prefix, keyMarker, uploadIDMarker string, maxUploads int) (resp api.MultipartListUploadsResponse, _ error) {
```

Comment:
```suggestion
func (s *SQLStore) ListMultipartUploads(ctx context.Context, bucket, prefix, keyMarker, uploadIDMarker string, maxUploads int) (resp api.MultipartListUploadsResponse, err error) {
```

might as well name the `error` and do `err = ... return`